File: frontend/src/api/paloalto.ts
import { apiClient } from "@/api/client";
import type { Paginated } from "@/types";

// Palo Alto（PAN-OS）整合。路徑帶 /api/v1 前綴（baseURL 為 /）。

export interface PaloAltoFirewall {
  id: string;
  name: string;
  api_url: string;
  enabled: boolean;
  verify_tls: boolean;
  /** REST URI 的版本段（如 v11.1）。留空＝自動偵測 */
  api_version: string | null;
  vsys_list: string[] | null;
  sync_dhcp: boolean;
  sync_arp: boolean;
  sync_policies: boolean;
  sync_nat: boolean;
  sync_addresses: boolean;
  sync_interval_seconds: number;
  description: string | null;
  scope_subnet_ids: string[] | null;
  last_sync_at: string | null;
  last_error: string | null;
}

export interface PaloAltoWrite {
  name: string;
  api_url: string;
  api_key?: string;
  enabled?: boolean;
  verify_tls?: boolean;
  api_version?: string | null;
  /** PATCH 的 null 是「不修改」→ 要改回自動偵測得用這個旗標 */
  clear_api_version?: boolean;
  vsys_list?: string[] | null;
  sync_dhcp?: boolean;
  sync_arp?: boolean;
  sync_policies?: boolean;
  sync_nat?: boolean;
  sync_addresses?: boolean;
  sync_interval_seconds?: number;
  description?: string;
  scope_subnet_ids?: string[];
}

/** 連線診斷：逐端點回報是否可讀與筆數（沒有實機時，這是唯一能對齊欄位的辦法） */
export interface PaloAltoDiagnosis {
  api_url: string;
  api_version: string;
  vsys: string[];
  ok_count: number;
  checks: { endpoint: string; ok: boolean; rows?: number; error?: string }[];
}

export interface PaloAltoPolicy {
  id: string; vsys: string; name: string; position: number;
  action: string | null; disabled: boolean;
  from_zone: string | null; to_zone: string | null;
  source: string | null; destination: string | null;
  application: string | null; service: string | null; description: string | null;
}

export interface PaloAltoAddressObject {
  id: string; vsys: string; name: string; kind: string;
  obj_type: string | null; value: string | null;
  members: string[] | null; description: string | null;
}

export async function listPaloAlto(): Promise<Paginated<PaloAltoFirewall>> {
  const { data } = await apiClient.get<Paginated<PaloAltoFirewall>>("/api/v1/paloalto", {
    params: { page: 1, page_size: 200 },
  });
  return data;
}

export async function createPaloAlto(p: PaloAltoWrite): Promise<PaloAltoFirewall> {
  const { data } = await apiClient.post<PaloAltoFirewall>("/api/v1/paloalto", p);
  return data;
}

export async function updatePaloAlto(
  id: string, p: Partial<PaloAltoWrite>,
): Promise<PaloAltoFirewall> {
  const { data } = await apiClient.patch<PaloAltoFirewall>(`/api/v1/paloalto/${id}`, p);
  return data;
}

export async function deletePaloAlto(id: string): Promise<void> {
  await apiClient.delete(`/api/v1/paloalto/${id}`);
}

export async function testPaloAlto(id: string): Promise<PaloAltoDiagnosis> {
  const { data } = await apiClient.post<PaloAltoDiagnosis>(`/api/v1/paloalto/${id}/test`);
  return data;
}

export async function syncPaloAlto(id: string): Promise<{ task_id: string }> {
  const { data } = await apiClient.post(`/api/v1/paloalto/${id}/sync`);
  return data;
}

export async function listPaloAltoPolicies(
  id: string, vsys?: string,
): Promise<PaloAltoPolicy[]> {
  const { data } = await apiClient.get<{ items: PaloAltoPolicy[] }>(
    `/api/v1/paloalto/${id}/policies`, { params: vsys ? { vsys } : undefined });
  return data.items ?? [];
}

export async function listPaloAltoAddresses(
  id: string, vsys?: string,
): Promise<PaloAltoAddressObject[]> {
  const { data } = await apiClient.get<{ items: PaloAltoAddressObject[] }>(
    `/api/v1/paloalto/${id}/addresses`, { params: vsys ? { vsys } : undefined });
  return data.items ?? [];
}

File: backend/app/schemas/paloalto.py
"""Palo Alto（PAN-OS）整合 schemas。"""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Annotated

from pydantic import Field, HttpUrl

from app.schemas.base import StrictModel

#: REST URI 裡的版本段。留空＝由 `show system info` 自行推導。
_VERSION_PATTERN = r"^v\d+\.\d+$"


class PaloAltoBase(StrictModel):
    name: Annotated[str, Field(min_length=1, max_length=128)]
    api_url: HttpUrl
    enabled: bool = True
    verify_tls: bool = True
    #: 留空＝自動偵測（強烈建議留空；填錯會讓所有 REST 端點回 404）
    api_version: Annotated[str | None, Field(pattern=_VERSION_PATTERN)] = None
    #: 留空＝自動探索全部 vsys
    vsys_list: list[Annotated[str, Field(min_length=1, max_length=64)]] | None = None
    sync_dhcp: bool = False
    sync_arp: bool = True
    sync_policies: bool = False
    sync_nat: bool = False
    sync_addresses: bool = False
    sync_interval_seconds: Annotated[int, Field(ge=30, le=86400)] = 300
    description: Annotated[str | None, Field(max_length=2048)] = None
    scope_subnet_ids: list[uuid.UUID] | None = None


class PaloAltoCreate(PaloAltoBase):
    api_key: Annotated[str, Field(min_length=1, max_length=1024)]


class PaloAltoUpdate(StrictModel):
    name: Annotated[str | None, Field(min_length=1, max_length=128)] = None
    api_url: HttpUrl | None = None
    api_key: Annotated[str | None, Field(min_length=1, max_length=1024)] = None
    enabled: bool | None = None
    verify_tls: bool | None = None
    api_version: Annotated[str | None, Field(pattern=_VERSION_PATTERN)] = None
    #: 明確清掉自訂版本、改回自動偵測（`None` 在 PATCH 是「不修改」）
    clear_api_version: bool = False
    vsys_list: list[Annotated[str, Field(min_length=1, max_length=64)]] | None = None
    sync_dhcp: bool | None = None
    sync_arp: bool | None = None
    sync_policies: bool | None = None
    sync_nat: bool | None = None
    sync_addresses: bool | None = None
    sync_interval_seconds: Annotated[int | None, Field(ge=30, le=86400)] = None
    description: Annotated[str | None, Field(max_length=2048)] = None
    scope_subnet_ids: list[uuid.UUID] | None = None


class PaloAltoRead(StrictModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    name: str
    api_url: str
    enabled: bool
    verify_tls: bool
    api_version: str | None = None
    vsys_list: list[str] | None = None
    sync_dhcp: bool
    sync_arp: bool
    sync_policies: bool
    sync_nat: bool
    sync_addresses: bool
    sync_interval_seconds: int
    description: str | None = None
    scope_subnet_ids: list[uuid.UUID] | None = None
    last_sync_at: datetime | None = None
    last_error: str | None = None
    created_at: datetime
    updated_at: datetime


class PaloAltoPolicyRead(StrictModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    vsys: str
    name: str
    position: int | None = None
    action: str | None = None
    disabled: bool | None = None
    from_zone: str | None = None
    to_zone: str | None = None
    source: str | None = None
    destination: str | None = None
    application: str | None = None
    service: str | None = None
    description: str | None = None
    last_sync_at: datetime | None = None


class PaloAltoAddressRead(StrictModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    vsys: str
    name: str
    obj_type: str | None = None
    kind: str
    value: str | None = None
    members: list | None = None
    description: str | None = None
    last_sync_at: datetime | None = None
